File: CG/MicroProject/lop.cpp
#include <graphics.h>
#include <conio.h>
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include <unistd.h>
const int COLOR_RED = COLOR(255, 51, 51);
const int COLOR_ORANGE = COLOR(255, 153, 51);
const int COLOR_YELLOW = COLOR(255, 255, 102);
const int COLOR_GREEN = COLOR(102, 255, 102);
const int COLOR_BLUE = COLOR(51, 153, 255);
const int COLOR_INDIGO = COLOR(102, 0, 204);
const int COLOR_VIOLET = COLOR(255, 102, 255);
const int WINDOW_WIDTH = 660;
const int WINDOW_HEIGHT = 550;
void drawDisk(const int &x, const int &y, const int &size) {
    static const int colors[] = {COLOR_RED, COLOR_ORANGE, COLOR_YELLOW, COLOR_GREEN, COLOR_BLUE, COLOR_INDIGO, COLOR_VIOLET};
    for (int i = 0; i < 20; ++i) {
        setfillstyle(SOLID_FILL, COLOR(
                                     GetRValue(colors[size - 1]) - i * 3,
                                     GetGValue(colors[size - 1]) - i * 3,
                                     GetBValue(colors[size - 1]) - i * 3));
        bar(x - size * 20, y + i, x + size * 20, y + i + 1);
    }
    setcolor(WHITE);
    setlinestyle(SOLID_LINE, 0, NORM_WIDTH);
    for (int i = 0; i < 3; ++i) {
        line(x - size * 20 + 5, y + 3 + i, x + size * 20 - 5, y + 3 + i);
        setcolor(COLOR(255, 255, 255));
    }
    setcolor(WHITE);
    for (int i = 0; i < 2; ++i) {
        rectangle(x - size * 20 - i, y - i, x + size * 20 + i, y + 20 + i);
    }
    for (int i = 0; i < 5; ++i) {
        line(x - size * 20 - 2, y - 2 + i, x - size * 20 + 5, y - 2 + i);
        line(x - size * 20 - 2 + i, y - 2, x - size * 20 - 2 + i, y + 5);
        line(x + size * 20 + 2, y - 2 + i, x + size * 20 - 5, y - 2 + i);
        line(x + size * 20 + 2 - i, y - 2, x + size * 20 + 2 - i, y + 5);
    }
}
inline int getTowerIndex(const int &x) {
    if (x < WINDOW_WIDTH / 3) return 0;
    if (x < 2 * WINDOW_WIDTH / 3) return 1;
    return 2;
}
int getTopDisk(int towers[3][7], const int &tower) {
    for (int i = 6; i >= 0; --i)
        if (towers[tower][i]) return towers[tower][i];
    return 0;
}
void drawTowers(int towers[3][7], const int &numDisks, char *moveText, const int &moveCount) {
    static const int towerX[3] = {WINDOW_WIDTH / 6, WINDOW_WIDTH / 2, 5 * WINDOW_WIDTH / 6};
    const int towerBaseHeight = WINDOW_HEIGHT - 140;
    const int btnX = WINDOW_WIDTH - 140;
    cleardevice();
    setfillstyle(SOLID_FILL, COLOR(10, 15, 35));
    bar(0, 0, WINDOW_WIDTH, WINDOW_HEIGHT);
    for (int i = 0; i < 100; ++i) {
        const int brightness = rand() % 256;
        putpixel(rand() % WINDOW_WIDTH, rand() % WINDOW_HEIGHT, COLOR(brightness, brightness, brightness));
    }
    int btnY = 12;
    setfillstyle(SOLID_FILL, COLOR(60, 179, 113));
    bar(btnX, btnY, btnX + 120, btnY + 30);
    for (int i = 0; i < 30; ++i) {
        setcolor(COLOR(60 + i, 179 + i / 2, 113 + i / 2));
        line(btnX, btnY + i, btnX + 120, btnY + i);
    }
    setcolor(WHITE);
    rectangle(btnX, btnY, btnX + 120, btnY + 30);
    rectangle(btnX - 1, btnY - 1, btnX + 121, btnY + 31);
    settextstyle(SANS_SERIF_FONT, HORIZ_DIR, 2);
    outtextxy(btnX + 10, btnY + 5, "Auto Solve");
    btnY += 40;
    setfillstyle(SOLID_FILL, COLOR(220, 60, 60));
    bar(btnX, btnY, btnX + 120, btnY + 30);
    for (int i = 0; i < 30; ++i) {
        setcolor(COLOR(220 + i / 2, 60 + i / 2, 60 + i / 2));
        line(btnX, btnY + i, btnX + 120, btnY + i);
    }
    setcolor(WHITE);
    rectangle(btnX, btnY, btnX + 120, btnY + 30);
    rectangle(btnX - 1, btnY - 1, btnX + 121, btnY + 31);
    settextstyle(SANS_SERIF_FONT, HORIZ_DIR, 2);
    outtextxy(btnX + 30, btnY + 5, "Reset");
    for (int i = 0; i < 3; ++i) {
        for (int j = 0; j < 25; ++j) {
            setfillstyle(SOLID_FILL, COLOR(180 + j * 2, 180 + j * 2, 180 + j * 2));
            bar(towerX[i] - 45, towerBaseHeight + j, towerX[i] + 45, towerBaseHeight + j + 1);
        }
        setcolor(COLOR(220, 220, 220));
        for (int b = 0; b < 3; ++b)
            rectangle(towerX[i] - 46 + b, towerBaseHeight - 1 + b, towerX[i] + 46 - b, towerBaseHeight + 26 - b);
        setfillstyle(SOLID_FILL, COLOR(200, 200, 200));
        bar(towerX[i] - 10, WINDOW_HEIGHT / 4, towerX[i] + 10, towerBaseHeight);
        setcolor(COLOR(240, 240, 240));
        line(towerX[i] - 10, WINDOW_HEIGHT / 4, towerX[i] - 10, towerBaseHeight);
        setcolor(COLOR(160, 160, 160));
        line(towerX[i] + 10, WINDOW_HEIGHT / 4, towerX[i] + 10, towerBaseHeight);
        setfillstyle(SOLID_FILL, COLOR(180, 180, 180));
        bar(towerX[i] - 20, WINDOW_HEIGHT / 4 - 15, towerX[i] + 20, WINDOW_HEIGHT / 4);
        for (int t = 0; t < 2; ++t) {
            setcolor(COLOR(220 - t * 30, 220 - t * 30, 220 - t * 30));
            rectangle(towerX[i] - 20 - t, WINDOW_HEIGHT / 4 - 15 - t, towerX[i] + 20 + t, WINDOW_HEIGHT / 4 + t);
        }
        for (int j = 0; j < numDisks; ++j)
            if (towers[i][j]) drawDisk(towerX[i], towerBaseHeight - 35 - j * 28, towers[i][j]);
        setcolor(WHITE);
        settextstyle(SANS_SERIF_FONT, HORIZ_DIR, 2);
        char label[15];
        sprintf(label, "Tower %c", 'A' + i);
        rectangle(towerX[i] - 40, towerBaseHeight + 35, towerX[i] + 40, towerBaseHeight + 60);
        rectangle(towerX[i] - 41, towerBaseHeight + 34, towerX[i] + 41, towerBaseHeight + 61);
        outtextxy(towerX[i] - 35, towerBaseHeight + 40, label);
    }
    setcolor(COLOR(255, 215, 0));
    settextstyle(SANS_SERIF_FONT, HORIZ_DIR, 2);
    rectangle(WINDOW_WIDTH / 2 - 200, 40, WINDOW_WIDTH / 2 + 200, 75);
    rectangle(WINDOW_WIDTH / 2 - 202, 38, WINDOW_WIDTH / 2 + 202, 77);
    outtextxy(WINDOW_WIDTH / 2 - 190, 50, moveText);
    char stepCountText[20];
    sprintf(stepCountText, "Moves: %d", moveCount);
    setcolor(WHITE);
    settextstyle(SANS_SERIF_FONT, HORIZ_DIR, 2);
    rectangle(10, 10, 120, 40);
    rectangle(8, 8, 122, 42);
    outtextxy(15, 15, stepCountText);
}
inline bool isMoveValid(int towers[3][7], const int &from, const int &to, const int &disk) {
    const int topDisk = getTopDisk(towers, to);
    return (topDisk == 0 || topDisk > disk);
}
void moveDisk(int towers[3][7], const int &from, const int &to, const int &disk) {
    for (int i = 0; i < 7; ++i) {
        if (towers[from][i] == disk) {
            towers[from][i] = 0;
            break;
        }
    }
    for (int i = 0; i < 7; ++i) {
        if (towers[to][i] == 0) {
            towers[to][i] = disk;
            break;
        }
    }
}
void solveTowerOfHanoi(int towers[3][7], const int &n, const int &from, const int &to, const int &aux, const int &numDisks, int &moveCount) {
    if (n == 1) {
        const int disk = getTopDisk(towers, from);
        moveDisk(towers, from, to, disk);
        ++moveCount;
        char msg[50];
        sprintf(msg, "Auto-moving disk from Tower %c to %c", 'A' + from, 'A' + to);
        drawTowers(towers, numDisks, msg, moveCount);
        delay(800);
        return;
    }
    solveTowerOfHanoi(towers, n - 1, from, aux, to, numDisks, moveCount);
    solveTowerOfHanoi(towers, 1, from, to, aux, numDisks, moveCount);
    solveTowerOfHanoi(towers, n - 1, aux, to, from, numDisks, moveCount);
}
int main() {
    int numDisks;
    printf("Enter number of disks (1-7): ");
    scanf("%d", &numDisks);
    if (numDisks < 1 || numDisks > 7) {
        printf("Invalid input! Please enter 1-7 disks.\n");
        return 1;
    }
    int gd = DETECT, gm;
    initgraph(&gd, &gm, "");
    int towers[3][7] = {{0}};
    for (int i = 0; i < numDisks; ++i) towers[0][i] = numDisks - i;
    int moveCount = 0;
    int selectedTower = -1;
    int x, y;
    drawTowers(towers, numDisks, "Click source and destination towers", moveCount);
    while (1) {
        if (ismouseclick(WM_LBUTTONDOWN)) {
            getmouseclick(WM_LBUTTONDOWN, x, y);
            if (x >= WINDOW_WIDTH - 140 && x <= WINDOW_WIDTH - 20 && y >= 12 && y <= 42) {
                memset(towers, 0, sizeof(towers));
                for (int i = 0; i < numDisks; ++i) towers[0][i] = numDisks - i;
                moveCount = 0;
                selectedTower = -1;
                solveTowerOfHanoi(towers, numDisks, 0, 2, 1, numDisks, moveCount);
                continue;
            }
            if (x >= WINDOW_WIDTH - 140 && x <= WINDOW_WIDTH - 20 && y >= 52 && y <= 82) {
                memset(towers, 0, sizeof(towers));
                for (int i = 0; i < numDisks; ++i) towers[0][i] = numDisks - i;
                moveCount = 0;
                selectedTower = -1;
                drawTowers(towers, numDisks, "Game Reset! Click source and destination towers", moveCount);
                continue;
            }
            const int clickedTower = getTowerIndex(x);
            if (selectedTower == -1) {
                const int topDisk = getTopDisk(towers, clickedTower);
                if (topDisk > 0) {
                    selectedTower = clickedTower;
                    char msg[50];
                    sprintf(msg, "Selected Tower %c", 'A' + clickedTower);
                    drawTowers(towers, numDisks, msg, moveCount);
                }
            } else {
                const int disk = getTopDisk(towers, selectedTower);
                if (isMoveValid(towers, selectedTower, clickedTower, disk)) {
                    moveDisk(towers, selectedTower, clickedTower, disk);
                    ++moveCount;
                    char msg[50];
                    sprintf(msg, "Moved disk from Tower %c to %c", 'A' + selectedTower, 'A' + clickedTower);
                    drawTowers(towers, numDisks, msg, moveCount);
                }
                selectedTower = -1;
            }
        }
        if (towers[2][numDisks - 1] != 0) {
            for (int i = 0; i < 50; ++i) {
                const int x = rand() % WINDOW_WIDTH;
                const int y = rand() % WINDOW_HEIGHT;
                const int color = rand() % 15 + 1;
                setcolor(color);
                circle(x, y, rand() % 5 + 1);
                delay(20);
            }
            char solvedText[50];
            sprintf(solvedText, "Puzzle Completed in %d Moves!", moveCount);
            setcolor(COLOR(255, 215, 0));
            settextstyle(SANS_SERIF_FONT, HORIZ_DIR, 3);
            rectangle(WINDOW_WIDTH / 2 - 250, 90, WINDOW_WIDTH / 2 + 250, 130);
            rectangle(WINDOW_WIDTH / 2 - 252, 88, WINDOW_WIDTH / 2 + 252, 132);
            outtextxy(WINDOW_WIDTH / 2 - 240, 100, solvedText);
            break;
        }
        delay(30);
    }
    getch();
    closegraph();
    return 0;
}

File: CG/MicroProject/lop.py
import pygame
import random
import time

# Modern color scheme
COLOR_RED = (231, 76, 60)      # Pomegranate
COLOR_ORANGE = (230, 126, 34)  # Carrot
COLOR_YELLOW = (241, 196, 15)  # Sun Flower  
COLOR_GREEN = (46, 204, 113)   # Emerald
COLOR_BLUE = (52, 152, 219)    # Peter River
COLOR_INDIGO = (155, 89, 182)  # Amethyst
COLOR_VIOLET = (142, 68, 173)  # Wisteria

# Dynamic window sizing based on screen resolution
display_info = pygame.display.Info()
WINDOW_WIDTH = int(display_info.current_w * 0.8)  # 80% of screen width
WINDOW_HEIGHT = int(WINDOW_WIDTH * 0.75)  # 4:3 aspect ratio

def draw_disk(screen, x, y, size):
    colors = [COLOR_RED, COLOR_ORANGE, COLOR_YELLOW, COLOR_GREEN, COLOR_BLUE, COLOR_INDIGO, COLOR_VIOLET]
    
    # Gradient effect
    for i in range(20):
        color = tuple(max(0, c - i*3) for c in colors[size-1])
        pygame.draw.rect(screen, color, (x - size*int(WINDOW_WIDTH/33), y + i, size*int(WINDOW_WIDTH/16.5), 1))
    
    # Metallic highlights
    for i in range(3):
        pygame.draw.line(screen, (255,255,255), (x - size*int(WINDOW_WIDTH/33) + 5, y + 3 + i), 
                        (x + size*int(WINDOW_WIDTH/33) - 5, y + 3 + i))
    
    # Border effect
    for i in range(2):
        pygame.draw.rect(screen, (255,255,255), 
                        (x - size*int(WINDOW_WIDTH/33) - i, y - i, size*int(WINDOW_WIDTH/16.5) + i*2, 20 + i*2), 1)
    
    # Corner details
    for i in range(5):
        pygame.draw.line(screen, (255,255,255), (x - size*int(WINDOW_WIDTH/33) - 2, y - 2 + i), 
                        (x - size*int(WINDOW_WIDTH/33) + 5, y - 2 + i))
        pygame.draw.line(screen, (255,255,255), (x - size*int(WINDOW_WIDTH/33) - 2 + i, y - 2), 
                        (x - size*int(WINDOW_WIDTH/33) - 2 + i, y + 5))
        pygame.draw.line(screen, (255,255,255), (x + size*int(WINDOW_WIDTH/33) + 2, y - 2 + i), 
                        (x + size*int(WINDOW_WIDTH/33) - 5, y - 2 + i))
        pygame.draw.line(screen, (255,255,255), (x + size*int(WINDOW_WIDTH/33) + 2 - i, y - 2), 
                        (x + size*int(WINDOW_WIDTH/33) + 2 - i, y + 5))

def get_tower_index(x):
    if x < WINDOW_WIDTH/3:
        return 0
    if x < 2*WINDOW_WIDTH/3:
        return 1
    return 2

def get_top_disk(towers, tower):
    for i in range(6, -1, -1):
        if towers[tower][i]:
            return towers[tower][i]
    return 0

def draw_towers(screen, towers, num_disks, move_text, move_count):
    tower_x = [WINDOW_WIDTH//6, WINDOW_WIDTH//2, 5*WINDOW_WIDTH//6]
    tower_base_height = int(WINDOW_HEIGHT * 0.75)
    btn_x = WINDOW_WIDTH - int(WINDOW_WIDTH * 0.2)
    
    # Modern dark background
    screen.fill((34, 47, 62))
    
    # Starfield effect
    for _ in range(int(WINDOW_WIDTH * 0.15)):
        brightness = random.randint(100, 255)
        pygame.draw.circle(screen, (brightness,brightness,brightness), 
                         (random.randint(0,WINDOW_WIDTH), random.randint(0,WINDOW_HEIGHT)), 1)
    
    # Buttons with modern styling
    btn_y = int(WINDOW_HEIGHT * 0.02)
    btn_width = int(WINDOW_WIDTH * 0.18)
    btn_height = int(WINDOW_HEIGHT * 0.055)
    
    # Auto Solve button
    pygame.draw.rect(screen, (46, 204, 113), (btn_x, btn_y, btn_width, btn_height))
    pygame.draw.rect(screen, (255,255,255), (btn_x, btn_y, btn_width, btn_height), 2)
    
    font = pygame.font.SysFont('Arial', int(WINDOW_HEIGHT * 0.044))
    text = font.render('Auto Solve', True, (255,255,255))
    screen.blit(text, (btn_x + btn_width*0.1, btn_y + btn_height*0.15))
    
    # Reset button
    btn_y += int(btn_height * 1.3)
    pygame.draw.rect(screen, (231, 76, 60), (btn_x, btn_y, btn_width, btn_height))
    pygame.draw.rect(screen, (255,255,255), (btn_x, btn_y, btn_width, btn_height), 2)
    
    text = font.render('Reset', True, (255,255,255))
    screen.blit(text, (btn_x + btn_width*0.25, btn_y + btn_height*0.15))
    
    # Draw towers
    for i in range(3):
        # Base
        pygame.draw.rect(screen, (189, 195, 199), 
                        (tower_x[i]-int(WINDOW_WIDTH*0.068), tower_base_height, 
                         int(WINDOW_WIDTH*0.136), int(WINDOW_HEIGHT*0.045)))
        
        # Pole
        pygame.draw.rect(screen, (189, 195, 199), 
                        (tower_x[i]-int(WINDOW_WIDTH*0.015), WINDOW_HEIGHT//4, 
                         int(WINDOW_WIDTH*0.03), tower_base_height-WINDOW_HEIGHT//4))
        
        # Top cap
        pygame.draw.rect(screen, (189, 195, 199), 
                        (tower_x[i]-int(WINDOW_WIDTH*0.03), WINDOW_HEIGHT//4-15, 
                         int(WINDOW_WIDTH*0.06), 15))
        
        # Draw disks
        for j in range(num_disks):
            if towers[i][j]:
                draw_disk(screen, tower_x[i], tower_base_height-35-j*28, towers[i][j])
        
        # Tower labels
        text = font.render(f'Tower {chr(65+i)}', True, (255,255,255))
        label_width = int(WINDOW_WIDTH * 0.12)
        pygame.draw.rect(screen, (255,255,255), 
                        (tower_x[i]-label_width//2, tower_base_height+35, label_width, btn_height), 1)
        screen.blit(text, (tower_x[i]-label_width//2+5, tower_base_height+40))
    
    # Move text
    text = font.render(move_text, True, (241, 196, 15))
    msg_width = int(WINDOW_WIDTH * 0.6)
    pygame.draw.rect(screen, (241, 196, 15), 
                    (WINDOW_WIDTH//2-msg_width//2, int(WINDOW_HEIGHT*0.07), msg_width, btn_height), 1)
    screen.blit(text, (WINDOW_WIDTH//2-msg_width//2+10, int(WINDOW_HEIGHT*0.08)))
    
    # Move counter
    text = font.render(f'Moves: {move_count}', True, (255,255,255))
    pygame.draw.rect(screen, (255,255,255), 
                    (10, 10, int(WINDOW_WIDTH*0.17), btn_height), 1)
    screen.blit(text, (15, 15))

def is_move_valid(towers, from_tower, to_tower, disk):
    top_disk = get_top_disk(towers, to_tower)
    return top_disk == 0 or top_disk > disk

def move_disk(towers, from_tower, to_tower, disk):
    for i in range(7):
        if towers[from_tower][i] == disk:
            towers[from_tower][i] = 0
            break
    
    for i in range(7):
        if towers[to_tower][i] == 0:
            towers[to_tower][i] = disk
            break

def solve_tower_of_hanoi(screen, towers, n, from_tower, to_tower, aux_tower, num_disks, move_count):
    if n == 1:
        disk = get_top_disk(towers, from_tower)
        move_disk(towers, from_tower, to_tower, disk)
        move_count[0] += 1
        msg = f"Auto-moving disk from Tower {chr(65+from_tower)} to {chr(65+to_tower)}"
        draw_towers(screen, towers, num_disks, msg, move_count[0])
        pygame.display.flip()
        time.sleep(0.8)
        return
    
    solve_tower_of_hanoi(screen, towers, n-1, from_tower, aux_tower, to_tower, num_disks, move_count)
    solve_tower_of_hanoi(screen, towers, 1, from_tower, to_tower, aux_tower, num_disks, move_count)
    solve_tower_of_hanoi(screen, towers, n-1, aux_tower, to_tower, from_tower, num_disks, move_count)

def main():
    while True:
        try:
            num_disks = int(input("Enter number of disks (1-7): "))
            if 1 <= num_disks <= 7:
                break
            print("Invalid input! Please enter 1-7 disks.")
        except ValueError:
            print("Please enter a valid number.")
    
    # Initialize pygame and check for errors
    if pygame.init()[1] > 0:
        print(f"Warning: {pygame.init()[1]} pygame modules failed to initialize")
    
    # Initialize video system specifically and check for errors
    if pygame.display.init():
        screen = pygame.display.set_mode((WINDOW_WIDTH, WINDOW_HEIGHT))
        pygame.display.set_caption("Tower of Hanoi")
    else:
        print("Error: Could not initialize pygame display")
        return
    
    towers = [[0]*7 for _ in range(3)]
    for i in range(num_disks):
        towers[0][i] = num_disks - i
    
    move_count = 0
    selected_tower = -1
    
    draw_towers(screen, towers, num_disks, "Click source and destination towers", move_count)
    pygame.display.flip()
    
    running = True
    while running:
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                running = False
            
            if event.type == pygame.MOUSEBUTTONDOWN and event.button == 1:
                x, y = event.pos
                btn_width = int(WINDOW_WIDTH * 0.18)
                btn_height = int(WINDOW_HEIGHT * 0.055)
                btn_x = WINDOW_WIDTH - int(WINDOW_WIDTH * 0.2)
                
                if btn_x <= x <= btn_x + btn_width:
                    if int(WINDOW_HEIGHT * 0.02) <= y <= int(WINDOW_HEIGHT * 0.02) + btn_height:
                        towers = [[0]*7 for _ in range(3)]
                        for i in range(num_disks):
                            towers[0][i] = num_disks - i
                        move_count = [0]
                        selected_tower = -1
                        solve_tower_of_hanoi(screen, towers, num_disks, 0, 2, 1, num_disks, move_count)
                        move_count = move_count[0]
                        continue
                    
                    if int(WINDOW_HEIGHT * 0.02) + btn_height*1.3 <= y <= int(WINDOW_HEIGHT * 0.02) + btn_height*2.3:
                        towers = [[0]*7 for _ in range(3)]
                        for i in range(num_disks):
                            towers[0][i] = num_disks - i
                        move_count = 0
                        selected_tower = -1
                        draw_towers(screen, towers, num_disks, "Game Reset! Click source and destination towers", move_count)
                        pygame.display.flip()
                        continue
                
                clicked_tower = get_tower_index(x)
                
                if selected_tower == -1:
                    top_disk = get_top_disk(towers, clicked_tower)
                    if top_disk > 0:
                        selected_tower = clicked_tower
                        msg = f"Selected Tower {chr(65+clicked_tower)}"
                        draw_towers(screen, towers, num_disks, msg, move_count)
                        pygame.display.flip()
                else:
                    disk = get_top_disk(towers, selected_tower)
                    if is_move_valid(towers, selected_tower, clicked_tower, disk):
                        move_disk(towers, selected_tower, clicked_tower, disk)
                        move_count += 1
                        msg = f"Moved disk from Tower {chr(65+selected_tower)} to {chr(65+clicked_tower)}"
                        draw_towers(screen, towers, num_disks, msg, move_count)
                        pygame.display.flip()
                    selected_tower = -1
        
        if towers[2][num_disks-1] != 0:
            # Victory celebration
            for _ in range(50):
                x = random.randint(0, WINDOW_WIDTH)
                y = random.randint(0, WINDOW_HEIGHT)
                color = (random.randint(100,255), random.randint(100,255), random.randint(100,255))
                pygame.draw.circle(screen, color, (x,y), random.randint(2,8))
                pygame.display.flip()
                time.sleep(0.02)
            
            font = pygame.font.SysFont('Arial', int(WINDOW_HEIGHT * 0.065))
            text = font.render(f"Puzzle Completed in {move_count} Moves!", True, (241, 196, 15))
            msg_width = int(WINDOW_WIDTH * 0.75)
            pygame.draw.rect(screen, (241, 196, 15), 
                           (WINDOW_WIDTH//2-msg_width//2, int(WINDOW_HEIGHT*0.16), msg_width, int(WINDOW_HEIGHT*0.073)), 1)
            screen.blit(text, (WINDOW_WIDTH//2-msg_width//2+10, int(WINDOW_HEIGHT*0.17)))
            pygame.display.flip()
            
            while True:
                for event in pygame.event.get():
                    if event.type == pygame.QUIT:
                        running = False
                        break
                if not running:
                    break
                time.sleep(0.03)
    
    pygame.quit()

if __name__ == "__main__":
    main()
